src/storage/mvcc/txn.rs

```diff
@@ -132,6 +141,169 @@ impl<S: Snapshot> MvccTxn<S> {
         Ok(self.reader.get_write(&key, ts)?.is_some())
     }
 
+    pub fn acquire_pessimistic_lock(
+        &mut self,
+        key: Key,
+        primary: &[u8],
+        for_update_ts: u64,
+        should_not_exist: bool,
+        options: &Options,
+    ) -> Result<()> {
+        if let Some(lock) = self.reader.load_lock(&key)? {
+            if lock.ts != self.start_ts {
+                return Err(Error::KeyIsLocked {
+                    key: key.into_raw()?,
+                    primary: lock.primary,
+                    ts: lock.ts,
+                    ttl: lock.ttl,
+                });
+            }
+            if lock.lock_type != LockType::Pessimistic {
+                return Err(Error::LockTypeNotMatch {
+                    start_ts: self.start_ts,
+                    key: key.into_raw()?,
+                    pessimistic: false,
+                });
+            }
+            MVCC_DUPLICATE_CMD_COUNTER_VEC.pessimistic_lock.inc();
+            return Ok(());
+        }
+
+        if let Some((commit_ts, write)) = self.reader.seek_write(&key, u64::max_value())? {
+            // The isolation level of pessimistic transactions is RC. `for_update_ts` is
+            // the commit_ts of the data this transaction read. If exists a commit version
+            // whose commit timestamp is larger than current `for_update_ts`, the
+            // transaction should retry to get the latest data.
+            if commit_ts > for_update_ts {
+                MVCC_CONFLICT_COUNTER.pessimistic_lock_conflict.inc();
+                return Err(Error::WriteConflict {
+                    start_ts: self.start_ts,
+                    conflict_start_ts: write.start_ts,
+                    conflict_commit_ts: commit_ts,
+                    key: key.into_raw()?,
+                    primary: primary.to_vec(),
+                });
+            }
+
+            // Handle rollback.
+            // If the start timestamp of write is equal to transaction's start timestamp
+            // as well as commit timestamp, the lock is already rollbacked.
+            if write.start_ts == self.start_ts && commit_ts == self.start_ts {
+                assert!(write.write_type == WriteType::Rollback);
+                return Err(Error::PessimisticLockRollbacked {
+                    start_ts: self.start_ts,
+                    key: key.into_raw()?,
+                });
+            }
+            // If `commit_ts` we seek is already before `start_ts`, the rollback must not exist.
+            if commit_ts > self.start_ts {
+                if let Some((commit_ts, write)) = self.reader.seek_write(&key, self.start_ts)? {
+                    if write.start_ts == self.start_ts {
+                        assert!(
+                            commit_ts == self.start_ts && write.write_type == WriteType::Rollback
+                        );
```

Comment:
What happens for the following case:
```
begin
select .. for update  //  k1 acquire lock?
update k1 = xxx;
commit   // what will k1 left?         
```

Reply:
If a key is already locked, it will not lock again in later statements.

Reply:
what do the comments on L198 mean
```
  // If `commit_ts` we seek is already before `start_ts`, the rollback must not exist.
         
```
while we still assert the write_type is Rollback here?

Reply:
If `commit_ts < start_ts`, we needn't to check `rollback`.